File: ExamplesTutorials/Tutorials/3-LatticeManipulation/LatticeManipulation.py
#!/usr/bin/env python

# Merlin++: C++ Class Library for Charged Particle Accelerator Simulations
# Copyright (c) 2001-2018 The Merlin++ developers
# This file is covered by the terms the GNU GPL version 2, or (at your option) any later version, see the file COPYING
# This file is derived from software bearing the copyright notice in merlin4_copyright.txt

from __future__ import division, print_function
import numpy
from matplotlib import pyplot
import os
import sys

data_filenames = ["tutorial3a.out","tutorial3b.out"]
dir_path = os.path.dirname(os.path.realpath(__file__))
print(dir_path)

data_paths = ["../../","../../../","../../../../","../../../../../"]
all_good = False;
for data_path in data_paths:
	try:
		fname = os.path.join(data_path, data_filenames[0])
		latticefunctions1 = numpy.loadtxt(fname)
		fname2 = os.path.join(data_path, data_filenames[1])
		latticefunctions2 = numpy.loadtxt(fname2)
		print("Read test data:", fname)
		print("Read test data:", fname2)
	except IOError:
		continue#

	all_good = True;
	pyplot.subplot(2,1,1)
	pyplot.ylabel("Beta x [m]")

	beta_x1, = pyplot.plot(latticefunctions1[:,0],latticefunctions1[:,7], color='b', label='Untouched')
	beta_x2, = pyplot.plot(latticefunctions2[:,0],latticefunctions2[:,7], color='b', linestyle='--', label='Manipulated')
	pyplot.legend(handles=[beta_x1, beta_x2],loc=1)

	pyplot.subplot(2,1,2)
	pyplot.ylabel("Beta y [m]")
	pyplot.xlabel("Location [m]")
	beta_x1, = pyplot.plot(latticefunctions1[:,0],latticefunctions1[:,9], color='r', label='Untouched')
	beta_x2, = pyplot.plot(latticefunctions2[:,0],latticefunctions2[:,9], color='r', linestyle='--', label='Manipulated')
	pyplot.legend(handles=[beta_x1, beta_x2],loc=1)


	pyplot.show()
	
if all_good:
	print("Plotted lattice functions")
	os.remove(data_filenames)
else:
	print("Failed to plot lattice functions. Could not read output file.")
	exit(1)

File: ExamplesTutorials/Tutorials/3-LatticeManipulation/LatticeManipulation.cpp
/*
 * Merlin++: C++ Class Library for Charged Particle Accelerator Simulations
 * Copyright (c) 2001-2018 The Merlin++ developers
 * This file is covered by the terms the GNU GPL version 2, or (at your option) any later version, see the file COPYING
 * This file is derived from software bearing the copyright notice in merlin4_copyright.txt
 */

//////////////////////////////////////////////////////////////////////////////////////////////////////////
//																										//
//	Tutorial 3 - LatticeManipulation																	//
//																										//
//	The following provides a simple example of how to manually manipulate the constructed lattice		//
//	in order to simulate alignment errors/changes														//
//																										//
//////////////////////////////////////////////////////////////////////////////////////////////////////////

// Include units
#include "PhysicalUnits.h"

// Include MAD interface classes
#include "MADInterface.h"

// Include lattice manipulation classes
#include "StandardMultipoles.h"
#include "MagnetMover.h"

// Include closed orbit calculator
#include "LatticeFunctions.h"
#include "ClosedOrbit.h"

// Namespaces for convenience
using namespace PhysicalUnits;
using namespace ParticleTracking;

// Define accelerator parameters for convenience
#define beamenergy 5.0*GeV

int main()
{
	cout << "Locating MAD lattice information..." << endl;
	// Loop over possible build directories to locate MAD .tfs files
	string paths[] = {"../input/StorageRing.tfs", "input/StorageRing.tfs", "Tutorials/input/StorageRing.tfs"};
	string lattice_path;
	for (size_t i=0; i<3; i++)
	{
		ifstream test_file;
		test_file.open(paths[i].c_str());
		if (test_file)
		{
			lattice_path = paths[i];
			break;
		}
	}
	cout << "MAD lattice information found at: " << lattice_path << endl;

	cout << "Constructing Lattice..." << endl;

	// Instantiate MADInterface with lattice file input
	MADInterface MADinput(lattice_path, beamenergy);

	// Construct Model
	AcceleratorModel* theModel = MADinput.ConstructModel();

	// Calculate beta and dispersion functions
	LatticeFunctionTable latticeFunctions(theModel,beamenergy);
	latticeFunctions.Calculate();

	// Write lattice functions to output file
	ofstream latticeFunctionLog("build/tutorial3a.out");
	latticeFunctions.PrintTable(latticeFunctionLog);

	// Manipulate lattice to simulation alignment errors etc
	vector<MagnetMover*> magnetMovers;
	theModel->ExtractTypedElements(magnetMovers);

	// Offset 20th magnet vertically by 100um
	magnetMovers[20]->SetY(100.0e-6);

	// Offset 40th magnet horizontally by 50um
	magnetMovers[40]->SetX(50.0e-6);

	// Extract a list of the quadrupoles
	vector<Quadrupole*> quadVec;
	theModel->ExtractTypedElements(quadVec);

	// Put a 20% gradient error on the 5th and 20th quadrupoles
	MultipoleField& field = quadVec[5]->GetField();
	Complex b1a = field.GetComponent(1);
	field.SetComponent(1, b1a.real() * 1.20, b1a.imag() * 1.20);

	MultipoleField& field2 = quadVec[20]->GetField();
	Complex b1b = field2.GetComponent(1);
	field2.SetComponent(1, b1b.real() * 1.10, b1b.imag() * 1.10);

	// Calculate beta and dispersion functions
	LatticeFunctionTable newLatticeFunctions(theModel,beamenergy);
	newLatticeFunctions.Calculate();

	// Write lattice functions to output file
	ofstream newLatticeFunctionLog("build/tutorial3b.out");
	newLatticeFunctions.PrintTable(newLatticeFunctionLog);

	delete theModel;

	cout << "Successful! Tutorial 3 Complete." << endl;
	cout << "Please run the corresponding python script to see a plot of manipulated lattice functions." << endl;

	return 0;
}
